test/ConstantProduct/DeploymentParamsTest.sol

```diff
@@ -6,13 +6,121 @@ import {Utils} from "test/libraries/Utils.sol";
+        IERC20 token0 = IERC20(Utils.addressFromString("DeploymentParamsTest: any token0"));
+        IERC20 token1 = IERC20(Utils.addressFromString("DeploymentParamsTest: any token1"));
+        mockSafeApprove(token0, expectedDeploymentAddress(), solutionSettler.vaultRelayer());
+        mockSafeApprove(token1, expectedDeploymentAddress(), solutionSettler.vaultRelayer());
+        mockSafeApprove(token0, expectedDeploymentAddress(), defaultDeployer());
+        mockSafeApprove(token1, expectedDeploymentAddress(), defaultDeployer());
+        vm.prank(defaultDeployer());
+        ConstantProduct constantProduct = new ConstantProduct(solutionSettler, token0, token1);
+        assertEq(address(constantProduct.solutionSettler()), address(solutionSettler));
+        assertEq(constantProduct.solutionSettlerDomainSeparator(), solutionSettler.domainSeparator());
+        assertEq(address(constantProduct.vaultRelayer()), address(solutionSettler.vaultRelayer()));
+        assertEq(address(constantProduct.token0()), address(token0));
+        assertEq(address(constantProduct.token1()), address(token1));
+    }
+
+    function approvedToken(string memory name) private returns (IERC20 token) {
+        token = IERC20(Utils.addressFromString(name));
+        mockSafeApprove(token, expectedDeploymentAddress(), solutionSettler.vaultRelayer());
+        mockSafeApprove(token, expectedDeploymentAddress(), defaultDeployer());
+        vm.mockCallRevert(address(token), hex"", abi.encode("Unexpected call to token contract"));
+    }
+
+    function revertingToken(string memory name, address spenderGoodApproval, address spenderBadApproval)
+        private
+        returns (IERC20 token)
+    {
+        token = IERC20(Utils.addressFromString(name));
+        mockSafeApprove(token, expectedDeploymentAddress(), spenderGoodApproval);
+        mockZeroAllowance(token, expectedDeploymentAddress(), spenderBadApproval);
+        vm.mockCallRevert(
+            address(token),
+            abi.encodeCall(IERC20.approve, (spenderBadApproval, type(uint256).max)),
+            "`approve` is an expected call"
+        );
+        vm.mockCallRevert(address(token), hex"", abi.encode("Unexpected call to token contract"));
+    }
+
+    function revertApproveVaultRelayerToken(string memory name) private returns (IERC20) {
+        return revertingToken(name, defaultDeployer(), solutionSettler.vaultRelayer());
+    }
+
+    function revertApproveDeployerToken(string memory name) private returns (IERC20) {
+        return revertingToken(name, solutionSettler.vaultRelayer(), defaultDeployer());
+    }
+
+    function testDeploymentAllowsVaultRelayerToken0() public {
+        IERC20 reverting = revertApproveVaultRelayerToken("reverting");
+        IERC20 regular = approvedToken("regular");
+        vm.expectRevert("`approve` is an expected call");
+        vm.prank(defaultDeployer());
+        new ConstantProduct(solutionSettler, reverting, regular);
```

Comment:
```suggestion
        new ConstantProduct(solutionSettler, regular, reverting);
```
Should these not be inverted, same as the test case for `Token1` below? (or if the idea is that the test is testing that `Token0` is being set - as verified by it failing, I think it may be better to make this clearer in the test name, or alternatively use `vm.expectCall` as I found this a logic not so straightforward on first read)

Reply:
I was convinced that `expectCall` was buggy and so I often do this workaround of catching reverts to make sure that something was called. But no, `expectCall` works exactly as you'd expect. I rewrote these tests to use that and I'll start using it much more often!